src/coprocessor/dag/expr/builtin_string.rs

```diff
@@ -517,6 +517,68 @@ impl ScalarFunc {
             return Ok(Some(Cow::Borrowed(b"")));
         }
     }
+
+    #[inline]
+    pub fn substring_binary_2_args<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        self.substring_binary(ctx, row, false)
+    }
+
+    #[inline]
+    pub fn substring_binary_3_args<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        self.substring_binary(ctx, row, true)
+    }
+
+    #[inline]
+    fn substring_binary<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+        with_len: bool,
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let s = try_opt!(self.children[0].eval_string(ctx, row));
+        let pos = try_opt!(self.children[1].eval_int(ctx, row));
+        let (len, len_positive) = if with_len {
+            let len = try_opt!(self.children[2].eval_int(ctx, row));
+            i64_to_usize(
+                len,
+                self.children[2]
+                    .field_type()
+                    .flag()
+                    .contains(FieldTypeFlag::UNSIGNED),
+            )
+        } else {
+            (s.len(), true)
+        };
+
+        if pos == 0 || len == 0 || !len_positive {
+            return Ok(Some(Cow::Borrowed(b"")));
+        }
+
+        let (pos, positive_search) = i64_to_usize(
+            pos,
+            self.children[1]
+                .field_type()
+                .flag()
+                .contains(FieldTypeFlag::UNSIGNED),
+        );
+
+        let start = if positive_search {
+            (pos - 1).min(s.len())
+        } else {
+            s.len().checked_sub(pos).unwrap_or_else(|| s.len())
+        };
+
+        let end = (start + len.min(s.len())).min(s.len());
```

Comment:
maybe `(start + len).min(s.len())` is enough?

Reply:
`start + len ` may overflow,  e.g: start = 2, len = u64::max_value()
there is a test case for it  https://github.com/tikv/tikv/blob/56733c60961608392a230e6a6de2048176674799/src/coprocessor/dag/expr/builtin_string.rs#L2008-L2013

Reply:
or maybe using `saturating_add` is better?
```
 let end = start.saturating_add(len).min(s.len());
```

Reply:
good idea! I prefer saturating_add.